R/layout.R

```diff
@@ -108,3 +112,312 @@ layout_column_wrap <- function(
+
+  children <- Map(f = bs_grid_wrapper, children, width_classes, fillable)
+
+  tag <- div(
+    class = "grid bslib-grid",
+    style = css(
+      "--bs-gap" = gap,
+      "--bs-columns" = n_cols,
+      # Always provide the `height:auto` default so that the CSS variable
+      # doesn't get inherited in a scenario like layout_column_wrap(height=200, ..., layout_column_wrap(...))
+      "--bslib-grid-height" = validateCssUnit(height %||% "auto"),
+      "--bslib-grid-height-mobile" = validateCssUnit(height_mobile %||% "auto"),
+      # For some reason, Bootstrap sets `grid-template-rows: 1fr` by default, which is
+      # problematic for a multi-row filling layout. This fixes it...
+      # > page_fillable(layout_columns(c(12, 12), plotly::plot_ly(), plotly::plot_ly()))
+      grid_auto_rows = "1fr"
+    ),
+    !!!attribs,
+    !!!children
+  )
+
+  tag <- bindFillRole(tag, item = fill)
+  tag <- tagAppendAttributes(tag, class = class)
+
+  as_fragment(
+    tag_require(tag, version = 5, caller = "layout_column_grid()")
+  )
+}
+
+bs_grid_wrapper <- function(el, bs_grid_classes = NULL, fillable = TRUE) {
+  div(
+    class = bs_grid_classes,
+    class = "bslib-gap-spacing",
+    if (fillable) as_fillable_container(),
+    el
+  )
+}
+
+bs_css_grid_width_classes <- function(breakpoints, n_kids, n_cols = 12) {
+  base_cols <- if (is_breakpoints(n_cols)) 12 else n_cols
+  classes <- as.list(rep(paste0("g-col-", base_cols), n_kids))
+
+  add_class <- function(idx, new) {
+    classes[[idx]] <<- c(classes[[idx]], new)
+  }
+
+  for (break_name in names(breakpoints)) {
+    bk <- breakpoints[[break_name]]
+    bk_cols <-
+      if (inherits(n_cols, "bslib_breakpoints")) {
+        n_cols[[break_name]]
+      } else {
+        n_cols
+      }
+
+    if (is.null(bk_cols)) {
+      if (any(is.na(bk$width))) {
+        # we can't infer the number of columns from the widths, so assume 12
+        bk_cols <- 12
+      } else {
+        bk_cols <- sum(bk$width + bk$before + bk$after)
+      }
+    }
+
+    if (length(bk$width) > n_kids) {
+      # TODO: more informative warning
+      rlang::warn("Too many `widths` provided; truncating")
+    }
+
+    widths <- rep_len(bk$width, n_kids)
+    before <- rep_len(bk$before, n_kids)
+    after <- rep_len(bk$after, n_kids)
+
+    # Fill NA widths with equal shares of remaining space
+    idx_na <- which(is.na(widths))
+    if (length(idx_na) > 0) {
+      n_accounted <- sum(widths, na.rm = TRUE) + sum(before) + sum(after)
+      n_remaining <- bk_cols - n_accounted
+      widths[idx_na] <- max(1, floor(n_remaining / length(idx_na)))
+    }
+
+
+    cursor <- 0L
+    update_cursor <- function(incr, can_split = FALSE) {
+      new <- cursor + incr
+      if (new > bk_cols) {
+        if (can_split) {
+          new <- new %% bk_cols
+        } else {
+          # signal that we've forced a move to the next row
+          new <- -1L
+        }
+      }
+      cursor <<- new
+    }
+
+    add_start_class <- FALSE
```

Comment:
Add some comments about what's going on (and maybe some visual ASCII descriptions)